### PR TITLE
Return NoProxy instead of DefaultProxy when url matches excludes list

### PR DESCRIPTION
**Pull request by troopa81**
## Description
Issue https://github.com/qgis/QGIS/issues/28034
When url was matching excludes list, returned proxy was the default one, meaning the one defined in the system. The returned one is now the NoProxy one.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by troopa81** [comment]
Please don't stale!

**Comment by nyalldawson** [comment]
This fix looks correct, but I have a slight concern in that the way this is exposed in the gui actually looks more like the current behavior. I.e. (to me) the UI suggests that any url entered in excluded urls just means "don't use the above proxy", not "don't use any proxy". So I'm concerned that someone may be relying on the current behavior, and this fix will break things for them.

I think we should expose this as a new setting, "Skip all proxies (including default system proxies) for URLs starting with"

**Comment by troopa81** [comment]
At the moment, you can choose a default proxy with an `excludes urls` list that in the end fall into the default proxy. That's a bit odd I think.

In firefox proxy configuration, excludes list is a `no proxy at all` list proxy

I would change the ui (replace "exclude urls" with "no proxy at all urls" and maybe add a no proxy in the combo box) rather than adding another settings. It will indeed break the configuration for people who use 2 proxies, one defined in QGIS and the system one. Is this really something that exist? People have normally only one proxy, no? If no, how could they manage to configure their other software (firefox for instance).

**Comment by haubourg** [comment]
@troopa81 +1 for the UI change, it looks much more natural this. Do we agree it is still a fix even if we slightly change the UI?

**Comment by nyalldawson** [comment]
Ok, I propose we go ahead and change the UI, BUT give a new settings key for the "no proxy at all"  URLS and leave the existing settings key and functionality untouched. And then add a shell example of the use of existing settings key to the `qgis_global_settings.ini` file (since it will no longer be exposed via the gui).

This way we don't risk breaking any existing setups.

**Comment by troopa81** [comment]
> Ok, I propose we go ahead and change the UI, BUT give a new settings key for the "no proxy at all" URLS and leave the existing settings key and functionality untouched.

Done 

> And then add a shell example of the use of existing settings key to the qgis_global_settings.ini file (since it will no longer be exposed via the gui).

Something like that 1ed9b20 ? Or you have something else in mind?

File: resources/qgis_global_settings.ini [hunk 1]
```diff
@@ -130,3 +130,8 @@ enable_problem_resolution=false
 [gps]
 # Default for GPS leap seconds correction as of 2019-06-19
 leapSecondsCorrection=18
+
+# [proxy]
+# # URL list for which proxy configuration doesn't apply. In the case of these URLs, systeme configuration
```
Comment (nyalldawson) [inline resources/qgis_global_settings.ini]:
```suggestion
# # URL list for which proxy configuration doesn't apply. In the case of these URLs, the default system proxy configuration
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Great - looks perfect to me!

**Pull request merged**